### PR TITLE
Color swatches

### LINKED ISSUE #3134: Color swatches

### PR DESCRIPTION
**Pull request by ruggi**
Fixes #3134 

**Problem:**

The color picker does not offer the ability to save colors that can be reused in the same project.

**Fix:**

This PR adds support for saving colors and reusing them inside the same project. The colors are available via the `Swatches` section in the color picker.

The swatches are saved using the hex color value for consistency.

The swatches color section has two actions:

1. A button to save the currently-selected color
2. A button to edit the available swatches; the scope is intentionally small, and when entering the "edit" mode, the only option is to delete a swatch by clicking on it.

In the future this can be arbitrarily extended with e.g. naming colors, reordering them, and so on.

Kapture 2023-01-19 at 11 56 14

### CODE REVIEW

File: editor/src/components/editor/store/store-hook-substore-helpers.ts [hunk 1]
```diff
@@ -173,4 +173,5 @@ export const EmptyEditorStateForKeysOnly: EditorState = {
   githubData: null as any,
   refreshingDependencies: false,
   assetChecksums: {},
+  colorSwatches: [],
```
Comment (balazsbajorics) [inline editor/src/components/editor/store/store-hook-substore-helpers.ts:176]: 👌

**Review by enidemi — APPROVED** [review]

File: editor/src/components/inspector/controls/color-picker-swatches.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  )
+
+  const [editing, setEditing] = React.useState(false)
+  const [colorSwatches, setColorSwatches] = React.useState<Array<ColorSwatch>>(storeColorSwatches)
+
+  const dispatch = useDispatch()
+
+  React.useEffect(() => {
+    dispatch([updateColorSwatches(colorSwatches)], 'everyone')
+  }, [colorSwatches, dispatch])
```
Comment (balazsbajorics) [inline editor/src/components/inspector/controls/color-picker-swatches.tsx:32]: interesting. I wonder if we should turn this into a hook that allows for "state with optimistic instant local update + global dispatch one frame later"

File: editor/src/components/inspector/controls/color-picker-swatches.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+          </Button>
+        )}
+      </FlexRow>
+
+      <FlexRow style={{ flexWrap: 'wrap', gap: 2 }}>
+        {colorSwatches.map((c, index) => {
+          return (
+            <Button
+              onClick={onClickPreset(c)}
+              key={index}
```
Comment (balazsbajorics) [inline editor/src/components/inspector/controls/color-picker-swatches.tsx:85]: index as key should be last resort, it's the same as not providing a key at all.

File: editor/src/components/inspector/controls/color-picker-swatches.tsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                zIndex: 0,
+              }}
+            >
+              {editing && <TrashIcon />}
+              {!editing && currentColor === c.hex && <CurrentColorBadge />}
+            </Button>
+          )
+        })}
+
+        {!editing && (
```
Comment (balazsbajorics) [inline editor/src/components/inspector/controls/color-picker-swatches.tsx:105]: we have the `unless` and `when` functions so we can avoid boolean operators in jsx

File: editor/src/components/inspector/controls/color-picker-swatches.tsx (hunk 3 above)
Comment (ruggi) [reply editor/src/components/inspector/controls/color-picker-swatches.tsx:85]: ah yeah good spot, it's a remnant of an early pass, fixing

File: editor/src/components/inspector/controls/color-picker-swatches.tsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        border: `1px solid ${colorTheme.border0.value}`,
+        borderRadius: '100%',
+        zIndex: 1,
+        boxShadow: `${colorTheme.canvasControlsSizeBoxShadowColor20.value} 0px 0px 1px, ${colorTheme.canvasControlsSizeBoxShadowColor21.value} 0px 1px 2px 1px`,
+      }}
+    />
+  )
+})
+
+const TrashIcon: React.FC = () => {
```
Comment (balazsbajorics) [inline editor/src/components/inspector/controls/color-picker-swatches.tsx:143]: minor but you can memoize TrashIcon and PlusIcon too

File: editor/src/components/inspector/controls/color-picker-swatches.tsx (hunk 3 above)
Comment (ruggi) [reply editor/src/components/inspector/controls/color-picker-swatches.tsx]: https://github.com/concrete-utopia/utopia/pull/3133/commits/22f1a93248cfe5a09cbb7f1b261e2974479a1111

File: editor/src/components/inspector/controls/color-picker-swatches.tsx (hunk 5 above)
Comment (ruggi) [reply editor/src/components/inspector/controls/color-picker-swatches.tsx]: https://github.com/concrete-utopia/utopia/pull/3133/commits/e56f43d4ed54256d7e1cb8fc0051b0b7566fb9c1

File: editor/src/components/inspector/controls/color-picker-swatches.tsx (hunk 4 above)
Comment (ruggi) [reply editor/src/components/inspector/controls/color-picker-swatches.tsx:107]: https://github.com/concrete-utopia/utopia/pull/3133/commits/e56f43d4ed54256d7e1cb8fc0051b0b7566fb9c1

Comment (balazsbajorics) [reply editor/src/components/inspector/controls/color-picker-swatches.tsx:107]: you missed line 107 :)

**Review by balazsbajorics — APPROVED** [review]

**Review by seanparsons — APPROVED** [review]

**Pull request merged**